### PR TITLE
Named instance binding support

### PR DESCRIPTION
**Pull request by Dominykas-Rapalis**
In this PR:

- Added a way to configure a multiple instances of a particular storage.

### CODE REVIEW

File: cloud-agnostic/core/src/NamedDependency.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import { Container, interfaces } from "inversify";
+
+import { Dependency } from "./Dependency";
+import { DependencyConfig } from "./DependencyConfig";
+import { ConfigError } from "./internal";
+
+export abstract class NamedDependency extends Dependency {
+  protected registerNamedInstance<TInstance>(
```
Comment (Animagne) [inline cloud-agnostic/core/src/NamedDependency.ts:12]: This assumes that there is only one class that will be resolved from the parent container and all the other classes in child container are there only to resolve it. That is the case for StorageDependency, but not this cloud agnostic framework in general. There could be multiple classes that you might want to resolve and they should share the same child container.

**Review by Animagne — CHANGES_REQUESTED** [review]
I think it would also be useful to have a sample for this specific use case, since it might come up for other users as well.

File: .vscode/launch.json [hunk 2]
```diff
@@ -9,16 +9,19 @@
...
-        "run",
-        "test:unit"
-      ],
-      "outFiles": [
-        "${workspaceFolder}/*/*/lib/**/*.js"
-      ],
-      "skipFiles": [
-        "<node_internals>/**"
-      ]
+      "runtimeArgs": ["run", "test:unit"],
```
[the remark this answers is not in the document]
Comment (Dominykas-Rapalis) [reply .vscode/launch.json]: Fixed

File: cloud-agnostic/core/src/NamedDependency.ts (hunk 1 above)
Comment (Dominykas-Rapalis) [reply cloud-agnostic/core/src/NamedDependency.ts]: Fixed

File: cloud-agnostic/core/src/NamedDependency.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import { Container, interfaces } from "inversify";
+
+import { Dependency } from "./Dependency";
+import { DependencyConfig } from "./DependencyConfig";
+
+export abstract class NamedDependency extends Dependency {
+  protected abstract bindInstances(
```
Comment (Animagne) [inline cloud-agnostic/core/src/NamedDependency.ts:11]: This name is also confusing. You're binding types for a single named instance, not for multiple named instances.

File: samples/src/client-file-download/NamedFileDownloader.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * Simple file downloader that depends on {@link ClientStorage}. Any specific
+ * implementation of {@link ClientStorage} can be passed to the constructor
+ * making this class cloud agnostic. {@link ClientStorage} instance is resolved
+ * using instance name (which is described in the configuration as instanceName property)
+ */
+@injectable()
+export class NamedFileDownloader {
+  constructor(
+    @inject(ClientStorage)
+    @named("instanceName")
```
Comment (Animagne) [inline samples/src/client-file-download/NamedFileDownloader.ts:19]: Would be good to have another class here, that would use a different instance name.

File: cloud-agnostic/core/src/NamedDependency.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    container: Container,
+    config: DependencyConfig
+  ): void {
+    const childContainer = container.createChild();
+
+    this.register(childContainer, config);
+    this.bindInstances(container, childContainer, config);
+  }
+
+  protected bindInstance<TInstance>(
```
[the remark this answers is not in the document]
Comment (Dominykas-Rapalis) [reply cloud-agnostic/core/src/NamedDependency.ts]: Fixed

File: cloud-agnostic/core/src/NamedDependency.ts (hunk 3 above)
Comment (Dominykas-Rapalis) [reply cloud-agnostic/core/src/NamedDependency.ts]: Fixed

File: samples/src/client-file-download/NamedFileDownloader.ts (hunk 4 above)
Comment (Dominykas-Rapalis) [reply samples/src/client-file-download/NamedFileDownloader.ts]: Fixed

**Review by Animagne — APPROVED** [review]

**Pull request merged**